language-plutus-core/test/Evaluation/CkMachine.hs

```diff
@@ -7,172 +7,91 @@ import           Language.PlutusCore
-import           Data.List
 import           Data.Text.Encoding                         (encodeUtf8)
 import           Test.Tasty
 import           Test.Tasty.Golden
 import           Test.Tasty.Hedgehog
 
-natToBool :: Quote (Type TyName ())
-natToBool = do
-    nat <- _recursiveType <$> getBuiltinNat
-    TyFun () nat <$> getBuiltinBool
-
-evenAndOdd :: Quote (Term TyName Name ())
-evenAndOdd = do
-    nat1 <- _recursiveType <$> getBuiltinNat
-    b1 <- getBuiltinBool
-    nat2 <- _recursiveType <$> getBuiltinNat
-    b2 <- getBuiltinBool
+getEvenAndOdd :: Quote (Tuple ())
+getEvenAndOdd = do
+    nat  <- _recursiveType <$> getBuiltinNat
+    bool <- getBuiltinBool
 
-    true <- getBuiltinTrue
+    true  <- getBuiltinTrue
     false <- getBuiltinFalse
 
-    fix2 <- getBuiltinFixN 2
-
-    q <- freshTyName () "Q"
-
-    choose <- freshName () "choose"
-    chooseTy <- do
-        ntb1 <- natToBool
-        ntb2 <- natToBool
-        pure $ TyFun () ntb1 (TyFun () ntb2 (TyVar () q))
-
     evenn <- freshName () "even"
-    evenTy <- natToBool
-    oddd <- freshName () "odd"
-    oddTy <- natToBool
+    oddd  <- freshName () "odd"
 
     let eoFunc b recc = do
           n <- freshName () "n"
-          nat <- _recursiveType <$> getBuiltinNat
-          bool <- getBuiltinBool
           pure $
               LamAbs () n nat $
               Apply () (Apply () (TyInst () (Unwrap () (Var () n)) bool) b) $ Var () recc
 
-    evenF <- eoFunc true oddd
-    oddF <- eoFunc false evenn
+    evenF <- Function () nat bool . Def evenn <$> eoFunc true oddd
+    oddF  <- Function () nat bool . Def oddd  <$> eoFunc false evenn
 
-    pure $
-        Apply () (foldl' (\acc t -> TyInst () acc t) fix2 [nat1, b1, nat2, b2]) $
-        TyAbs () q (Type ()) $
-        LamAbs () choose chooseTy $
-        LamAbs () evenn evenTy $
-        LamAbs () oddd oddTy $
-        Apply () (Apply () (Var () choose) evenF) oddF
+    getBuiltinMutualFixOf () [evenF, oddF]
 
 getEven :: Quote (Term TyName Name ())
-getEven = do
-    both <- evenAndOdd
-    instTy <- natToBool
-    x <- freshName () "x"
-    xTy <- natToBool
-    y <- freshName () "y"
-    yTy <- natToBool
-    pure $
-        Apply () (TyInst () both instTy) $
-        LamAbs () x xTy $
-        LamAbs () y yTy $
-        Var () x
-
-listInt :: Quote (Type TyName ())
-listInt = do
-    list1 <- _recursiveType <$> getBuiltinList
-    nat1 <- _recursiveType <$> getBuiltinNat
-    pure $ TyApp () list1 nat1
-
-listIntToListInt :: Quote (Type TyName ())
-listIntToListInt = TyFun () <$> listInt <*> listInt
-
-evenAndOddList :: Quote (Term TyName Name ())
-evenAndOddList = do
-    fix2 <- getBuiltinFixN 2
-
-    q <- freshTyName () "Q"
-
-    choose <- freshName () "choose"
-    chooseTy <- do
-        li1 <- listIntToListInt
-        li2 <- listIntToListInt
-        pure $ TyFun () li1 (TyFun () li2 (TyVar () q))
+getEven = getEvenAndOdd >>= tupleTermAt () 0
```

Comment:
This *is* nice, but I think it's better to have abstraction in the generated code too. It does make accessing it more awkward, I admit. But I think we could still do something like this, but just generate the appropriately instantiated tuple accessor.

Reply:
> This is nice, but I think it's better to have abstraction in the generated code too.

For sure. It was just a stupid bug that caused tuple duplication.

> But I think we could still do something like this, but just generate the appropriately instantiated tuple accessor.

An application of an accessor still duplicates a tuple. What matters duplication-wise is whether a tuple is a variable or really an inlined tuple. Whether you define an accessor and then apply it to a tuple or just apply the tuple directly to a selector is irrelevant, because you're going to duplicate the tuple in any case (which is fine if the tuple is a variable).

Reply:
Right, fair.